packages/cdktf-cli/src/lib/local-provider-versions.ts

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) HashiCorp, Inc
+// SPDX-License-Identifier: MPL-2.0
+import fs from "fs-extra";
+import path from "path";
+import { CdktfConfig } from "./cdktf-config";
+import { Errors } from "./errors";
+import { logger } from "./logging";
+
+/**
+ * Class to help with reading `versions.json` file
+ * published by `provider get` or `provider add` commands
+ */
+export class LocalProviderVersions {
+  private versions: Record<string, string> | undefined;
+
+  private get jsonFilePath(): string {
+    const config = CdktfConfig.read();
+    return path.resolve(config.codeMakerOutput, "versions.json");
+  }
+
+  private async readLocalVersionsJson(): Promise<void> {
+    let versionsJson;
+    try {
+      versionsJson = await fs.readFile(this.jsonFilePath, "utf8");
+    } catch (e) {
+      // If we cannot read the versions.json file, it may not exist
+      logger.debug("versions.json file reading error: ", e);
+      this.versions = undefined;
+      return;
+    }
+
+    let providerVersions;
+    try {
+      providerVersions = JSON.parse(versionsJson) as LocalProviderVersions;
+    } catch (e) {
+      throw Errors.External(
+        "versions.json file is malformed. The root must be a JSON object."
+      );
+    }
+
+    this.versions = Object.fromEntries(
```

Comment:
Fun fact: @ansgarm is going to love this :D I seriously don't understand how this is more readable than

```ts
Object.entries(providerVersions).reduce((obj, [providerFqn, versions]) => ({
   ...obj,
   [providerFqn
            .replace("registry.terraform.io/", "")
            .replace("hashicorp/", "")
            .toLowerCase()]: versions
}), {})
```

Reply:
Haha, I've been told to stay away from `reduce` too 😅 . I haven't used fromEntries before, and I agree I don't like the wrapper, but honestly: both approaches are ugly. So, I'll do the one that at least one person prefers 😇

Reply:
Haha, now you're stuck in some kind of "who's your favorite child" situation because Daniel and I each like the opposite one 😂

Reply:
I could see me 🤝 on the "both are ugly" though 😁

Reply:
oh no. Time to go back to lodash 😎